external-crates/move/move-compiler/src/sui_mode/typing.rs

```diff
@@ -179,15 +189,90 @@ impl<'a> TypingVisitorContext for Context<'a> {
     }
 }
 
-//**********************************************************************************************
+//**************************************************************************************************
+// Structs
+//**************************************************************************************************
+
+fn struct_def(context: &mut Context, name: StructName, sdef: &N::StructDefinition) {
+    const KEY_MSG: &str = "The 'key' ability is used to declare objects in Sui";
+    let N::StructDefinition {
+        warning_filter: _,
+        index: _,
+        attributes: _,
+        abilities,
+        type_parameters: _,
+        fields,
+    } = sdef;
+    let Some(key_loc) = abilities.ability_loc_(Ability_::Key) else {
+        // not an object, no extra rules
+        return;
+    };
+
+    let StructFields::Defined(fields) = fields else { return };
+    let invalid_first_field = if fields.is_empty() {
+        // no fields
+        Some(name.loc())
+    } else {
+        let (first_field_loc, first_field_name, _) =
+            fields.iter().find(|(_, _, (idx, _))| *idx == 0).unwrap();
+        if *first_field_name != ID_FIELD_NAME {
+            Some(first_field_loc)
+        } else {
+            None
+        }
+    };
+    if let Some(loc) = invalid_first_field {
```

Comment:
I'd be tempted to abstract this error out to take the location as an argument, and then write the above code more directly:

```
if fields.is_empty() {
   produce_error(name.loc())
   return;
}
let Some((lfield_oc, first_field_name, (_, first_field_type))) = fields.iter.find(|(_, _, (idx, _))| *idx == 0) else {
  // ICE?
}
if *first_field_name != ID_FIELD_NAME {
   produce_error(field_loc)
   return;
}
... <rest of the code>
```